processing/src/main/java/org/apache/druid/frame/read/columnar/StringArrayFrameColumnReader.java

```diff
@@ -0,0 +1,186 @@
+import org.apache.druid.query.rowsandcols.column.ColumnAccessorBasedColumn;
+import org.apache.druid.segment.ColumnValueSelector;
+import org.apache.druid.segment.column.BaseColumn;
+import org.apache.druid.segment.column.ColumnCapabilitiesImpl;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.data.ReadableOffset;
+
+/**
+ * Reader for {@link ColumnType#STRING_ARRAY}.
+ */
+public class StringArrayFrameColumnReader implements FrameColumnReader
+{
+  final int columnNumber;
+
+  /**
+   * Create a new reader.
+   *
+   * @param columnNumber column number
+   */
+  StringArrayFrameColumnReader(int columnNumber)
+  {
+    this.columnNumber = columnNumber;
+  }
+
+  @Override
+  public Column readRACColumn(Frame frame)
+  {
+    final Memory memory = frame.region(columnNumber);
+    validate(memory);
+
+    // String arrays always store multiple values per row.
+    assert isMultiValue(memory);
+    final long positionOfLengths = getStartOfStringLengthSection(frame.numRows());
+    final long positionOfPayloads = getStartOfStringDataSection(memory, frame.numRows());
+
+    StringFrameColumnReader.StringFrameColumn frameCol = new StringFrameColumnReader.StringFrameColumn(
+        frame,
+        true,
+        memory,
+        positionOfLengths,
+        positionOfPayloads,
+        true
+    );
+
+    return new ColumnAccessorBasedColumn(frameCol);
+  }
+
+  @Override
+  public ColumnPlus readColumn(final Frame frame)
+  {
+    final Memory memory = frame.region(columnNumber);
+    validate(memory);
+
+    // String arrays always store multiple values per row.
+    assert isMultiValue(memory);
+    final long startOfStringLengthSection = getStartOfStringLengthSection(frame.numRows());
+    final long startOfStringDataSection = getStartOfStringDataSection(memory, frame.numRows());
+
+    final BaseColumn baseColumn = new StringArrayFrameColumn(
+        frame,
+        memory,
+        startOfStringLengthSection,
+        startOfStringDataSection
+    );
+
+    return new ColumnPlus(
+        baseColumn,
+        new ColumnCapabilitiesImpl().setType(ColumnType.STRING_ARRAY)
+                                    .setHasMultipleValues(false)
+                                    .setDictionaryEncoded(false),
+        frame.numRows()
+    );
+  }
+
+  private void validate(final Memory region)
+  {
+    // Check if column is big enough for a header
+    if (region.getCapacity() < StringFrameColumnWriter.DATA_OFFSET) {
+      throw DruidException.defensive("Column[%s] is not big enough for a header", columnNumber);
+    }
+
+    final byte typeCode = region.getByte(0);
+    if (typeCode != FrameColumnWriters.TYPE_STRING_ARRAY) {
+      throw DruidException.defensive(
+          "Column[%s] does not have the correct type code; expected[%s], got[%s]",
+          columnNumber,
+          FrameColumnWriters.TYPE_STRING_ARRAY,
+          typeCode
+      );
+    }
+  }
+
+  static boolean isMultiValue(final Memory memory)
```

Comment:
nit: this method name leaves a bad taste in my mouth, string arrays have nothing to do with multi-value strings, for example `ColumnCapabilities.hasMultipleValues` is _never_ true for an array type, so I think we should avoid confusing terminology if we really are going to split this out. Why do we even need to check anything? Array writers shouldn't need to write multiple different things into the same frame

Reply:
removed the explicit check and only validate the type